src/core/mac/mac_frame.cpp

```diff
@@ -1165,16 +1165,20 @@ otError TxFrame::GenerateEnhAck(const RxFrame &aFrame, bool aIsFramePending, con
         SetDstAddr(address);
     }
 
-    // Set security header
-    SuccessOrExit(error = aFrame.GetSecurityControlField(securityControlField));
-    SuccessOrExit(error = aFrame.GetFrameCounter(frameCounter));
-    SuccessOrExit(error = aFrame.GetKeyId(keyId));
-
     SetPsduLength(kMaxPsduSize); // At this time the length of ACK hasn't been determined, set it to
                                  // `kMaxPsduSize` to call methods that check frame length.
-    SetSecurityControlField(securityControlField);
-    SetFrameCounter(frameCounter);
-    SetKeyId(keyId);
+
+    // Set security header
+    if (aFrame.GetSecurityEnabled())
+    {
+        SuccessOrExit(error = aFrame.GetSecurityControlField(securityControlField));
+        SuccessOrExit(error = aFrame.GetFrameCounter(frameCounter));
```

Comment:
For enh-ack, the sequence number should be the one from received frame. But the Auxiliary header frame counter should be increased monotonously. Any reason here you use the counter from received frame? I think it breaks the criterial that each key material should be used only once per frame.

Reply:
It should use the device's own frame counter. I will change it after #4976 gets merged since it updates a new way to get the frame counter.

Reply:
Thanks. Let's fix this after #4976 got merged.